### PR TITLE
Feature/osi 3.0.0

### PR DESCRIPTION
**Pull request by yichi-lu**

### CODE REVIEW

**Comment by joshw** [comment]
Yichi-  I'd like to review this, but I will wait until you resolve the conflicts.  You may want to split the api client changes into a separate PR for simplicity.

**Comment by pcarinhas** [comment]
Will the version number be kept at 2.1.2 at release or will it be 2.2.0?

**Comment by joshw** [comment]
I generally refer to http://semver.org/ on versioning, and i'd interpret that to mean that this would be 2.2.0, since it adds functionality in a backwards-compatible manner.

**Comment by pcarinhas** [comment]
Migrate scripts need to bump version to 2.2.0. Tested and works with modeler when doing so.

File: ZenPacks/zenoss/OpenStackInfrastructure/Instance.py [hunk 1]
```diff
@@ -47,7 +47,7 @@ def set_host_name(self, hostname):
         if len(hypervisors):
             hypervisor = sorted(hypervisors)[0]
             if len(hypervisors) > 1:
-                LOG.error("Multiple hypervisors were found matching hostname %s - Choosing %s" %
+                LOG.warn("Multiple hypervisors were found matching hostname %s - Choosing %s" %
```
Comment (joshw) [inline ZenPacks/zenoss/OpenStackInfrastructure/Instance.py]: Why the change?   This is something that should not be happening, and it's likely to cause problems if it is- that's why it was an error. 

Were you seeing this occur?

File: ZenPacks/zenoss/OpenStackInfrastructure/apiclients/txapiclient.py [hunk 2]
```diff
@@ -0,0 +1,839 @@
+###########################################################################
+#
+# This program is part of Zenoss Core, an open source monitoring platform.
+# Copyright (C) 2014, Zenoss Inc.
```
Comment (joshw) [inline ZenPacks/zenoss/OpenStackInfrastructure/apiclients/txapiclient.py]: Fix copyright date

File: ZenPacks/zenoss/OpenStackInfrastructure/interfaces.py [hunk 3]
```diff
@@ -145,3 +145,122 @@ def getNeutronIntegrationKeys(self):
...
+        objects.
+        """
+
+    """
+    Objects that implement this interface must also invoke
+    the "index_cinder_implementation_object" and "unindex_cinder_implementation_object"
+    functions provided in ZenPacks.zenoss.OpenStackInfrastructure.cinder_integration
+    from their index_object and unindex_object methods, respectively.
+
+    An example may be found in "README.cinder_integration.txt"
```
Comment (joshw) [inline ZenPacks/zenoss/OpenStackInfrastructure/interfaces.py]: README file is missing.  You need to either add it, or rename the neutron one and rewrite it to cover both.

File: ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py [hunk 4]
```diff
@@ -111,46 +103,47 @@ def collect(self, device, unused):
...
 
-        result = yield client.servers(detailed=True, search_opts={'all_tenants': 1})
+        # get hypervisor details for each individual hypervisor
+        results['hypervisor_details'] = {}
+        for hypervisor in results['hypervisors']:
+            result = yield client.nova_hypervisors(
+                hypervisor_id=hypervisor['id'])
+            hypervisor_id = prepId("hypervisor-{0}".format(hypervisor['id']))
+            results['hypervisor_details'][hypervisor_id] = result['hypervisor']
+
```
Comment (joshw) [inline ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py]: What is going on here?    I see no explanation for this change.  Why are you making all these extra API calls?

I see the comment below:    "if results['hypervisors_detailed'] did not give us hypervisor type, hypervisor version, try results['hypervisors_details']"

Need some more explanation of why you're incurring extra API calls here- If it is necessary on some versions, then I would suggest making it conditional, so that this extra stuff isn't done all the time.

Changes like this really should be tied to a specific jira with more explanation, since this changes existing behavior.

File: ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py [hunk 5]
```diff
@@ -239,6 +232,9 @@ def collect(self, device, unused):
         except Exception, e:
             log.warning("Unable to determine nova URL for nova-api component discovery: %s" % e)
 
+        # not being able to resolve hostname could, in some cases,
+        # result in nova-api and/or cinder-api components not being modeled
+        # make sure /etc/hosts file on the controller host is decent
```
Comment (joshw) [inline ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py]: Can you explain what you mean by this?  I wouldn't think the hosts file on the openstack controller would be important.

File: ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py [hunk 6]
```diff
@@ -263,16 +259,48 @@ def collect(self, device, unused):
...
+        result = yield client.cinder_volumesnapshots()
+        results['snapshots'] = result['snapshots']
+
+        # try:
+        #     result = yield client.cinder_volumebackups()
+        # except (BadRequestError, UnauthorizedError, NotFoundError, APIClientError) as e:
+        #     log.info('Server (not zenpack) error while calling client.cinder_volumebackups(): %s\n' % e.message)
+        #     results['backups'] = []
+        # else:
+        #     results['backups'] = result['backups']
```
Comment (joshw) [inline ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py]: Is this disabled for a reason?   Can/should it be removed?

File: ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py [hunk 7]
```diff
@@ -311,12 +339,6 @@ def process(self, device, results, unused):
             if not image.get('id', None):
                 continue
 
-            # If it's a snapshot, rather than a normal image, ignore it for
-            # the time being.
-            if 'server' in image:
-                log.debug("Ignoring image %s" % image.get('name', ''))
-                continue
-
```
Comment (joshw) [inline ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py]: Looking into this, I think you appear to be missing missing a relationship between snapshot and instance.

I think you were attempting to address that by counting snapshots as images (and thus relating them to instances), but you don't relate the images to snapshot components, so that doesn't work.  And even if it did, it would be combining two things- the image (snapshot) that an instance was built from, and the instance that the snapshot was built from.

So I think you need to add a direct snapshot to instance relationship (which instance is this a snapshot of), assuming openstack tracks that.

The other issue, that of instances that are built from snapshots rather than traditional instances.. I would suggest we need to talk about that a bit more.  Can you revert this change for the time being and we revisit that topic?  I am concerned that adding every snapshot to the instance list is going to bloat the system considerably with what are mostly duplicate components.  

I think the right solution to that may be somewhat complex and we need to talk it over, but that can be addressed in a separate pull request.

File: ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py [hunk 8]
```diff
@@ -416,7 +433,7 @@ def process(self, device, results, unused):
                     set_image=prepId('image-{0}'.format(image_id)),       # image-346eeba5-a122-42f1-94e7-06cb3c53f690
                     set_tenant=prepId('tenant-{0}'.format(tenant_id)),    # tenant-a3a2901f2fd14f808401863e3628a858
                     hostId=server.get('hostId', ''),                      # a84303c0021aa53c7e749cbbbfac265f
-                    hostName=server.get('name', '')                       # cloudserver01
+                    hostName=server.get('OS-EXT-SRV-ATTR:hypervisor_hostname', '')                       # cloudserver01
```
Comment (joshw) [inline ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py]: Whoah.  You're changing the meaning of this attribute.  It is supposed to be the hostname of the instance, not of the hypervisor host!

File: ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py [hunk 9]
```diff
@@ -504,6 +562,10 @@ def process(self, device, results, unused):
         hosts = []
         for host_id in hostmap:
             data = hostmap[host_id]
+
+            if not isValidHostname(data.get('hostname', host_id)):
```
Comment (joshw) [inline ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py]: What is it if it's not a hostname?   Should it at least log a message about this invalid data that it's ignoring?

File: ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py [hunk 10]
```diff
@@ -544,8 +622,26 @@ def process(self, device, results, unused):
                 title='{0}.{1}'.format(hypervisor.get('hypervisor_hostname', ''),
                                        hypervisor['id']),
                 hypervisorId=hypervisor['id'],  # 1
-                hypervisor_type=hypervisor_type.get(hypervisor_id, None),
+                hypervisor_type=hypervisor_type.get(hypervisor_id, '').upper(),
                 hypervisor_version=hypervisor_version.get(hypervisor_id, None),
+                host_ip=results['hypervisor_details'].get(hypervisor_id,
```
Comment (joshw) [inline ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py]: This appears to be the explanation of your other change.  Where did the requirement for all this come from?

I think this hypervisor modeling stuff needs to go into a separate pull request, it's not related to the cinder work and I don't know where the requirement came from.  

It's possible that this may not work in the vmware hypervisor scenario.  It would at least need need to be tested there, and I would rather not hold up this pull request on that.

File: ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py [hunk 11]
```diff
@@ -596,10 +698,10 @@ def process(self, device, results, unused):
         for hostname in sorted(nova_api_hosts):
             title = '{0}@{1} ({2})'.format('nova-api', hostname, device.zOpenStackRegionName)
             host_id = prepId("host-{0}".format(hostname))
-            nova_api_id = prepId('service-nova-api-{0}-{1}'.format(service.get('host', ''), device.zOpenStackRegionName))
+            nova_api_id = prepId('service-nova-api-{0}-{1}'.format(hostname, device.zOpenStackRegionName))
 
             services.append(ObjectMap(
-                modname='ZenPacks.zenoss.OpenStackInfrastructure.NovaApi',
+                modname='ZenPacks.zenoss.OpenStackInfrastructure.NovaService',
```
Comment (joshw) [inline ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py]: Explain?  Why is NovaApi a NovaService now?

File: ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py [hunk 12]
```diff
@@ -800,6 +924,139 @@ def process(self, device, results, unused):
...
+                title=volume.get('name', ''),
+                volumeId=volume['id'],  # 847424
+                avzone=volume.get('availability_zone', ''),
+                created_at=volume.get('created_at', '').replace('T', ' '),
+                sourceVolumeId=volume.get('source_volid', ''),
+                host=volume.get('os-vol-host-attr:host', ''),
+                size=volume.get('size', 0),
+                bootable=volume.get('bootable', 'FALSE').upper(),
+                status=volume.get('status', 'UNKNOWN').upper(),
+                set_tenant=prepId('tenant-{0}'.format(volume.get('os-vol-tenant-attr:tenant_id',''))),
```
Comment (joshw) [inline ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py]: This leaves it with "tenant-" as the tenant if there is no os-vol-tenant-attr:tenant_id.   Not really a great default- it will result in errors in the zenhub log with no real way to fix them..  I would omit set_tenant if there's no tenant id.

File: ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py [hunk 13]
```diff
@@ -800,6 +924,139 @@ def process(self, device, results, unused):
...
+                modname='ZenPacks.zenoss.OpenStackInfrastructure.Snapshot',
+                data=dict(
+                    id=prepId('snapshot-{0}'.format(snapshot['id'])),
+                    title=snapshot.get('name', ''),
+                    created_at=snapshot.get('created_at', '').replace('T', ' '),
+                    size=snapshot.get('size', 0),
+                    description=snapshot.get('description', ''),
+                    status=snapshot.get('status', 'UNKNOWN').upper(),
+                    set_volume=prepId('volume-{0}'.format(snapshot.get('volume_id',''))),
+                    set_tenant=prepId('tenant-{0}'.format(snapshot.get('os-extended-snapshot-attributes:project_id',''))),
```
Comment (joshw) [inline ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py]: Same concern for this one.  I would think we can safely assume that volume_id will be present, but any of these extended attributes might be missing, depending on the backend.   Better to avoid the relationship than to set it to a bogus value, which will fail.

File: ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py [hunk 14]
```diff
@@ -800,6 +924,139 @@ def process(self, device, results, unused):
...
+        #             )))
+        #
+        # Pools
+        pools = []
+        for pool in results['volume_pools']:
+            # does pool have id?
+
+            # pool name from Ceph looks like: 'block1@ceph#ceph
+            # the right most (or the middle part?) part is volume type.
+            # the middle part is Ceph cluster name?
```
Comment (joshw) [inline ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py]: If these questions have been answered, remove the comments.

File: ZenPacks/zenoss/OpenStackInfrastructure/Instance.py (hunk 1 above)
Comment (yichi-lu) [inline ZenPacks/zenoss/OpenStackInfrastructure/Instance.py]: I saw this happening quite often, and usually no bad things happen even if we have this 'error' msg. That's why I changed it to warning.

File: ZenPacks/zenoss/OpenStackInfrastructure/utils.py [hunk 15]
```diff
@@ -313,4 +313,14 @@ def getNetSubnetsGws_from_GwInfo(external_gateway_info):
 
     return (network, subnets, gateways)
 
+def isValidHostname(hostname):
```
Comment (joshw) [inline ZenPacks/zenoss/OpenStackInfrastructure/utils.py]: This seems unnecessarily strict to me.  I'm not really clear on why we need it at all.  (see previous comment)

File: ZenPacks/zenoss/OpenStackInfrastructure/zenpack.yaml [hunk 16]
```diff
@@ -13,6 +13,9 @@ class_relationships:
 # Containing Relations
 - Endpoint(components) 1:MC OpenstackComponent
 - Instance 1:MC Vnic
+# Cinder
+- Volume 1:MC Snapshot
```
Comment (joshw) [inline ZenPacks/zenoss/OpenStackInfrastructure/zenpack.yaml]: Are you positive that snapshots can not and will never move between volumes?

Containing relationships are dangerous  - I usually avoid them completely unless they are necessary because the object has no unique ID without its parent (which may be the case here, I don't know)

File: ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py (hunk 5 above)
Comment (yichi-lu) [inline ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py]: this is a comment for developers. I should change /etc/hosts to something like 'dns be able to resolve controller'

File: ZenPacks/zenoss/OpenStackInfrastructure/zenpack.yaml [hunk 17]
```diff
@@ -53,12 +62,14 @@ class_relationships:
...
 - Hypervisor 1:1 Host
+# Cinder
+- Instance 1:M Volume
 device_classes: !ZenPackSpec
   /Server/SSH/Linux/NovaHost:
     zProperties:
       zCollectorPlugins: [zenoss.cmd.uname, zenoss.cmd.uname_a, zenoss.cmd.df, zenoss.cmd.linux.cpuinfo,
         zenoss.cmd.linux.memory, zenoss.cmd.linux.ifconfig, zenoss.cmd.linux.netstat_an,
-        zenoss.cmd.linux.netstat_rn, zenoss.cmd.linux.process, zenoss.cmd.linux.rpm,
+        zenoss.cmd.linux.netstat_rn, zenoss.cmd.linux.process, zenoss.cmd.linux.lvm, zenoss.cmd.linux.rpm,
```
Comment (joshw) [inline ZenPacks/zenoss/OpenStackInfrastructure/zenpack.yaml]: I believe you are introducing a dependency on an unreleased version of LinuxMonitor here.   We would need to add that dependency to setup.py.  But really, I think what this is pointing out is that we shouldn't be setting zCollectorPlugins this way in the first place.  

Can you open a jira to track this issue?  We really want to just append our plugins on top of whatever ones already exist, not to hardcode the list.  We can do this in python code if necessary.   

I'm ok with merging this as it is for now, but we need to fix this before we can release.

File: ZenPacks/zenoss/OpenStackInfrastructure/zenpack.yaml [hunk 18]
```diff
@@ -408,50 +422,119 @@ device_classes: !ZenPackSpec
...
+                dpName: openstack_poolQoSSupportCount
+                format: '%7.0lf'
       Instance:
         description: 'OpenStackInfrastructure: Instance Monitoring'
         targetPythonClass: ZenPacks.zenoss.OpenStackInfrastructure.Instance
         datasources:
           cpuUtilization:
-            type: OpenStack Ceilometer Perf AMQP
+#            type: OpenStack Ceilometer Perf AMQP
+            type: OpenStack Ceilometer Perf API
```
Comment (joshw) [inline ZenPacks/zenoss/OpenStackInfrastructure/zenpack.yaml]: Huge red flag here.

We can not switch from AMQP to the API at this time.  Please revert these changes.

File: ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py (hunk 6 above)
Comment (yichi-lu) [inline ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py]: We have not worked on backup yet. it should be disabled. I'll fix that.

File: ZenPacks/zenoss/OpenStackInfrastructure/zenpack.yaml [hunk 19]
```diff
@@ -936,22 +1051,6 @@ classes: !ZenPackSpec
     dynamicview_relations:
       impacted_by: [hostedOn]
       impacts: [orgComponent]
-  NovaApi:
```
Comment (joshw) [inline ZenPacks/zenoss/OpenStackInfrastructure/zenpack.yaml]: As I mentioned before, why is this changing?   It is not a service (it doesn't show in the nova service list, and can't be monitored in that way, and probably has different impact relationships) -  it is its own kind of software component.

File: ZenPacks/zenoss/OpenStackInfrastructure/zenpacklib.py [hunk 20]
```diff
@@ -27,7 +27,7 @@
 """
 
 # PEP-396 version. (https://www.python.org/dev/peps/pep-0396/)
-__version__ = "1.1.0dev"
+__version__ = "1.0.3"
```
Comment (joshw) [inline ZenPacks/zenoss/OpenStackInfrastructure/zenpacklib.py]: This looks wrong- are you reverting to an *older* version of ZPL?

File: impact.yuml [hunk 21]
```diff
@@ -51,3 +54,20 @@
 # // [note: Cisco UCS;ZenPack{bg:darkslateblue}]-[UCS Service Profile{bg:darkslateblue}]
 [UCS Service Profile{bg:darkslateblue}]1-*>[Linux;Device{bg:deepskyblue}]
 
+<<<<<<< Updated upstream
```
Comment (joshw) [inline impact.yuml]: merge conflict here- need to clean that up

File: ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py (hunk 7 above)
Comment (yichi-lu) [inline ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py]: Let's discuss this a bit more

File: ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py (hunk 9 above)
Comment (yichi-lu) [inline ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py]: I'll add a log msg

File: ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py (hunk 11 above)
Comment (yichi-lu) [inline ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py]: Did you mention at one point we can merge nova api and nova service together?

File: ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py (hunk 12 above)
Comment (yichi-lu) [inline ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py]: if a tenant does not have an id, then all bets are off.

File: impact.yuml [hunk 22]
```diff
@@ -51,3 +54,20 @@
...
+=======
+# Cinder
+[CinderAgent{bg:orange}]*-*>[Volume{bg:orange}]
```
Comment (joshw) [inline impact.yuml]: There doesn't appear to be a component in your model called a cinder agent.  You may need to enhance the model, or change the way the impact works.

I would also be concerned about the cardinality on this- if there is an agent of some sort on every host, you don't want to get into a situation where if any host goes down, every volume is impacted.

File: ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py (hunk 13 above)
Comment (yichi-lu) [inline ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py]: I'll add some guard around it

File: impact.yuml [hunk 23]
```diff
@@ -51,3 +54,20 @@
...
+[CinderAgent{bg:orange}]*-*>[Snapshot{bg:orange}]
+[CinderAgent{bg:orange}]*-*>[Backup{bg:orange}]
+[Volume{bg:orange}]*-1>[Tenant{bg:orange}]
+[Snapshot{bg:orange}]*-1>[Tenant{bg:orange}]
+[Backup{bg:orange}]*-1>[Tenant{bg:orange}]
+[Quota{bg:orange}]1-1>[Tenant{bg:orange}]
+[Instance{bg:orange}]<1-1[Volume{bg:orange}]
+[VolumeType{bg:orange}]<1-*[Volume{bg:orange}]
```
Comment (joshw) [inline impact.yuml]: VolumeType doesn't sound like a thing that should be involved in impact.

File: impact.yuml [hunk 24]
```diff
@@ -51,3 +54,20 @@
 # // [note: Cisco UCS;ZenPack{bg:darkslateblue}]-[UCS Service Profile{bg:darkslateblue}]
 [UCS Service Profile{bg:darkslateblue}]1-*>[Linux;Device{bg:deepskyblue}]
 
+<<<<<<< Updated upstream
+=======
+# Cinder
+[CinderAgent{bg:orange}]*-*>[Volume{bg:orange}]
+[CinderAgent{bg:orange}]*-*>[Snapshot{bg:orange}]
+[CinderAgent{bg:orange}]*-*>[Backup{bg:orange}]
+[Volume{bg:orange}]*-1>[Tenant{bg:orange}]
```
Comment (joshw) [inline impact.yuml]: I see no monitoring of volumes.

File: impact.yuml [hunk 25]
```diff
@@ -51,3 +54,20 @@
...
+[Snapshot{bg:orange}]*-1>[Tenant{bg:orange}]
```
Comment (joshw) [inline impact.yuml]: I see no monitoring of snapshots, quota, etc.

File: impact.yuml [hunk 26]
```diff
@@ -51,3 +54,20 @@
...
+[Backup{bg:orange}]*-1>[Tenant{bg:orange}]
```
Comment (joshw) [inline impact.yuml]: Backup modeling is disabled.

File: ZenPacks/zenoss/OpenStackInfrastructure/apiclients/txapiclient.py [hunk 27]
```diff
@@ -0,0 +1,839 @@
...
+    """Wrapper for HTTP 401 Unauthorized error."""
+    pass
+
+
+class NotFoundError(APIClientError):
+    """Wrapper for HTTP 400 Bad Request error."""
+    pass
+
+
+@inlineCallbacks
```
Comment (joshw) [inline ZenPacks/zenoss/OpenStackInfrastructure/apiclients/txapiclient.py]: It would be nice to see some version of this turned into a unit test.  I do recognize that is challenging.

File: ZenPacks/zenoss/OpenStackInfrastructure/Instance.py (hunk 1 above)
Comment (joshw) [inline ZenPacks/zenoss/OpenStackInfrastructure/Instance.py]: Please log a jira with details, i think it needs investigation.  Multiple hypervisors with the same hostname, at least on a standard boring configuration of openstack with libvirt, doesn't seem normal, and if it's happening, we could be losing data.

File: ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py (hunk 11 above)
Comment (joshw) [inline ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py]: I think it might be possible to do, but i don't think it's this easy.  More investigation is required.

**Comment by joshw** [comment]
It also appears that that incremental model updates (via ceilometer events) are not implemented at this time.  Is that not possible, or just not done yet?

File: ZenPacks/zenoss/OpenStackInfrastructure/utils.py (hunk 15 above)
Comment (yichi-lu) [inline ZenPacks/zenoss/OpenStackInfrastructure/utils.py]: Jake Lampe had a customer with bogus hostname (something like xxx.yyy.com:xxxxx), and he kept asking Phil and me about why modeler would not work. So I thought we would verify for decent hostname

Comment (joshw) [inline ZenPacks/zenoss/OpenStackInfrastructure/utils.py]: Ignoring it altogether doesn't seem productive...

File: ZenPacks/zenoss/OpenStackInfrastructure/zenpack.yaml (hunk 16 above)
Comment (yichi-lu) [inline ZenPacks/zenoss/OpenStackInfrastructure/zenpack.yaml]: I am not aware of moving snapshots between volumes use case, at least within openstack. In openstack, a snapshot has a volume attribute, pointing back to its original volume.

File: ZenPacks/zenoss/OpenStackInfrastructure/zenpacklib.py (hunk 20 above)
Comment (yichi-lu) [inline ZenPacks/zenoss/OpenStackInfrastructure/zenpacklib.py]: I don't know what happened. I'd like to use the latest

File: impact.yuml (hunk 22 above)
Comment (yichi-lu) [inline impact.yuml]: I need to work on this

File: impact.yuml (hunk 23 above)
Comment (yichi-lu) [inline impact.yuml]: correct

File: impact.yuml (hunk 24 above)
Comment (yichi-lu) [inline impact.yuml]: please ignore impact.yuml right now. That;s why I say cinder is not ready yet

File: ZenPacks/zenoss/OpenStackInfrastructure/apiclients/txapiclient.py (hunk 27 above)
Comment (yichi-lu) [inline ZenPacks/zenoss/OpenStackInfrastructure/apiclients/txapiclient.py]: yes, we need to have some unit tests for twisted api client. we just need a mock api server somewhere

Comment (yichi-lu) [inline ZenPacks/zenoss/OpenStackInfrastructure/apiclients/txapiclient.py]: incremental modeling not done yet.

File: ZenPacks/zenoss/OpenStackInfrastructure/interfaces.py (hunk 3 above)
Comment (yichi-lu) [inline ZenPacks/zenoss/OpenStackInfrastructure/interfaces.py]: I'll fix it.

File: ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py (hunk 8 above)
Comment (yichi-lu) [inline ZenPacks/zenoss/OpenStackInfrastructure/modeler/plugins/zenoss/OpenStackInfrastructure.py]: my mistake

**Comment by pcarinhas** [comment]
Can we remove these two errors:
modeler/plugins/zenoss/OpenStackInfrastructure.py|48 col 1 error| 'BadRequestError' imported but unused [F401]               
modeler/plugins/zenoss/OpenStackInfrastructure.py|48 col 1 error| 'UnauthorizedError' imported but unused [F401]

**Comment by pcarinhas** [comment]
Might as well fix formatting in *NEW* code:

modeler/plugins/zenoss/OpenStackInfrastructure.py|159 col 74 error| the backslash is redundant between brackets [E502]
modeler/plugins/zenoss/OpenStackInfrastructure.py|190 col 74 error| the backslash is redundant between brackets [E502]
modeler/plugins/zenoss/OpenStackInfrastructure.py|274 col 9 error| too many blank lines (2) [E303]
modeler/plugins/zenoss/OpenStackInfrastructure.py|531 col 9 error| too many blank lines (2) [E303]
modeler/plugins/zenoss/OpenStackInfrastructure.py|810 col 24 error| missing whitespace around operator [E225]
modeler/plugins/zenoss/OpenStackInfrastructure.py|823 col 78 error| the backslash is redundant between brackets [E502]
modeler/plugins/zenoss/OpenStackInfrastructure.py|896 col 24 error| missing whitespace around operator [E225]
modeler/plugins/zenoss/OpenStackInfrastructure.py|922 col 9 error| too many blank lines (2) [E303]
modeler/plugins/zenoss/OpenStackInfrastructure.py|928 col 25 error| missing whitespace around operator [E225]
modeler/plugins/zenoss/OpenStackInfrastructure.py|981 col 31 error| missing whitespace around operator [E225]
modeler/plugins/zenoss/OpenStackInfrastructure.py|986 col 26 error| missing whitespace around operator [E225]
modeler/plugins/zenoss/OpenStackInfrastructure.py|991 col 27 error| missing whitespace around operator [E225]
modeler/plugins/zenoss/OpenStackInfrastructure.py|1050 col 1 error| continuation line missing indentation or outdented [E122]

**Comment by pcarinhas** [comment]
Fix formatting in new code (this is our only chance to fix it):

apiclients/txapiclient.py|14 col 1 error| 'sys' imported but unused [F401]                                                                          
apiclients/txapiclient.py|37 col 1 error| expected 2 blank lines, found 1 [E302]
apiclients/txapiclient.py|79 col 5 error| too many blank lines (2) [E303]       
apiclients/txapiclient.py|87 col 5 error| too many blank lines (2) [E303]       
apiclients/txapiclient.py|95 col 5 error| too many blank lines (2) [E303]       
apiclients/txapiclient.py|103 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|111 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|119 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|125 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|131 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|138 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|144 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|150 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|156 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|162 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|168 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|174 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|180 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|186 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|192 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|198 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|204 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|210 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|216 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|222 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|228 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|234 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|240 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|246 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|252 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|258 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|264 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|270 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|276 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|282 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|288 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|294 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|300 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|306 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|312 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|497 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|504 col 5 error| too many blank lines (2) [E303]      
apiclients/txapiclient.py|568 col 18 warning| .has_key() is deprecated, use 'in' [W601]
apiclients/txapiclient.py|577 col 1 error| expected 2 blank lines, found 1 [E302]
apiclients/txapiclient.py|581 col 1 error| expected 2 blank lines, found 1 [E302]

**Comment by pcarinhas** [comment]
Should we fix the extra import?:
datasources/PerfCeilometerAPIDataSource.py:14:1: F401 'ast' imported but unused

**Comment by pcarinhas** [comment]
Errors in *NEW* tests/test_moddler:
tests/test_model.py:24:1: F401 'ZVersion' imported but unused
tests/test_model.py:26:1: F401 'IInfo' imported but unused
tests/test_model.py:27:1: F401 'DeviceProxy' imported but unused
tests/test_model.py:54:9: E303 too many blank lines (2)
tests/test_model.py:67:5: E303 too many blank lines (2)
tests/test_model.py:217:73: E502 the backslash is redundant between brackets
tests/test_model.py:226:74: E502 the backslash is redundant between brackets
tests/test_model.py:257:75: E502 the backslash is redundant between brackets
tests/test_model.py:258:82: E502 the backslash is redundant between brackets
tests/test_model.py:297:1: E302 expected 2 blank lines, found 1
